api/protocol.go

```diff
@@ -52,68 +50,148 @@ type Response struct {
 	Data    map[string]interface{} `json:"data,omitempty"`
 }
 
-// ReadMessage reads a message from reader. A message is either a Request or a
-// Response
-func ReadMessage(reader io.Reader, msg interface{}) error {
-	buf := make([]byte, headerLength)
-	n, err := reader.Read(buf)
+func maxOpcodeForFrameType(t FrameType) int {
+	switch t {
+	default:
+		fallthrough
+	case TypeCommand:
+		return int(CmdMax)
+	case TypeResponse:
+		return int(CmdMax)
+	case TypeStream:
+		return int(StreamMax)
+	case TypeNotification:
+		return int(NotificationMax)
+	}
+}
+
+// ReadFrame reads a full frame (header and payload) from r.
+func ReadFrame(r io.Reader) (*Frame, error) {
+	// Read the header.
+	buf := make([]byte, minHeaderLength)
+	n, err := r.Read(buf)
 	if err != nil {
-		return err
+		return nil, err
 	}
-	if n != headerLength {
-		return errors.New("couldn't read the full header")
+	if n != minHeaderLength {
+		return nil, errors.New("frame: couldn't read the full header")
 	}
 
-	header := header{
-		length: binary.BigEndian.Uint32(buf[0:4]),
-		flags:  binary.BigEndian.Uint32(buf[4:8]),
+	// Decode it.
+	frame := &Frame{}
+	header := &frame.Header
+	header.Version = int(binary.BigEndian.Uint16(buf[0:2]))
+	if header.Version < 2 || header.Version > Version {
+		return nil, fmt.Errorf("frame: bad version %d", header.Version)
+	}
+	header.HeaderLength = int(buf[2]) * 4
+	header.Type = FrameType(buf[6] & 0xf)
+	flags := buf[6] & 0xf0
```

Comment:
flagInError is 0x10. Note that the "reserved" bits are actuallly specified as MBZ (must be zero), the sentence in the doc says: "Reserved fields are reserved for future use and must be zeroed." 

You're saying you are OR'ing with 0x10, I'm guessing you mean AND'ing?

In the code above we're AND'ing with 0xf0 to retrieve the 4 bits of flags, 3 MBZ bits and flagInError. That's what I intended indeed. 

I'm not sure how the proxy managed to output 0010 0001, are you sure of that? It's quite possible to have the tests working but not being coherent with the documentation though. 

The write path, what you receive, is:

```
	flags := byte(0)
 	if frame.Header.InError {
 		flags |= flagInError
	}
	buf[6] = flags | byte(header.Type)&0xf
```

`&` has a higher precedence than `|` (https://golang.org/ref/spec), and flagInError should go through that unchanged. So, puzzling!